### PR TITLE
Add CIL lexer/filetype

### PR DESCRIPTION
**Pull request by killerbees19**
Closes: #3305

* Common Intermediate Language: https://en.wikipedia.org/wiki/Common_Intermediate_Language
* Upstream discussion: https://sourceforge.net/p/scintilla/feature-requests/1265/

This PR adds a new Lexer (`CIL`) from upstream v5.1.4 and adds a new filetype in Geany. Thanks to @xv for his great work with the Lexer module! 🥳

I've tried to follow the instructions from `HACKING` file, thanks to @elextr for pointing me in the right direction. Sadly I've no experience with Geany and Lexilla in terms of coding internals. If something looks wrong, I'll try to fix the PR.

### CODE REVIEW

File: scintilla/lexilla/src/Lexilla.cxx [hunk 1]
```diff
@@ -178,6 +178,7 @@ static void AddGeanyLexers()
 		&lmBash,
 		&lmBatch,
 		&lmCaml,
+		&lmCIL,
```
Comment (elextr) [inline scintilla/lexilla/src/Lexilla.cxx:181]: `AddGeanyLexers()` is not part of Lexilla, it is added when Lexilla is updated by `scintilla_changes.patch` so you need to add there as well, or you lexer will be unknown and the crash will return on the next Lexilla update.

Comment (killerbees19) [reply scintilla/lexilla/src/Lexilla.cxx:181]: Thanks!

Fixed: 1397da184180e39a2797ffeb9f15c5174f1f52c5

File: data/filedefs/filetypes.cil [hunk 2]
```diff
@@ -0,0 +1,36 @@
+# For complete documentation of this file, please see Geany's main documentation
+[styling]
+# Edit these in the colorscheme .conf file instead
+default=default
+comment=comment
+commentline=comment_line
+stringliteral=string
```
Comment (techee) [inline data/filedefs/filetypes.cil:7]: I'd suggest renaming `stringliteral` to `string` here.

File: scintilla/scintilla_changes.patch [hunk 3]
```diff
@@ -62,7 +62,7 @@ diff --git scintilla/lexilla/src/Lexilla.cxx scintilla/lexilla/src/Lexilla.cxx
 index cd4b23617..af4a73db4 100644
 --- scintilla/lexilla/src/Lexilla.cxx
 +++ scintilla/lexilla/src/Lexilla.cxx
-@@ -165,12 +165,69 @@ namespace {
+@@ -165,12 +165,71 @@ namespace {
```
Comment (techee) [inline scintilla/scintilla_changes.patch:65]: Drop this diff - the patch should succeed without it.

File: src/highlighting.c [hunk 4]
```diff
@@ -1567,6 +1569,9 @@ gboolean highlighting_is_string_style(gint lexer, gint style)
 				style == SCE_H_SGML_SIMPLESTRING ||
 				style == SCE_H_SINGLESTRING);
 
+		case SCLEX_CIL:
+			return (style == SCE_CIL_STRING);
```
Comment (techee) [inline src/highlighting.c:1573]: I'd also suggest adding SCE_CIL_STRINGEOL here.

File: src/highlightingmappings.h [hunk 5]
```diff
@@ -336,6 +336,31 @@ static const HLKeyword highlighting_keywords_CAML[] =
...
+#define highlighting_lexer_CIL		SCLEX_CIL
+static const HLStyle highlighting_styles_CIL[] =
+{
+	{ SCE_CIL_DEFAULT,		"default",			FALSE },
+	{ SCE_CIL_COMMENT,		"comment",			FALSE },
+	{ SCE_CIL_COMMENTLINE,	"commentline",		FALSE },
+	{ SCE_CIL_WORD,			"keyword",			FALSE },
+	{ SCE_CIL_WORD2,		"keyword2",			FALSE },
+	{ SCE_CIL_WORD3,		"keyword3",			FALSE },
+	{ SCE_CIL_STRING,		"stringliteral",	FALSE },
```
Comment (techee) [inline src/highlightingmappings.h:349]: I'd suggest renaming this to plain "string" and updating `filetypes.cil` accordingly.

**Review by techee — CHANGES_REQUESTED** [review]
@killerbees19 Thanks for all your work and your patience! I'll try to assist you to get this merged to Geany finally.

A few more general comments apart from those commenting specific code:
1. It would be good to rebase this pull request on top of current master.
2. If there's an updated CIL lexer in Lexilla, please use it (but from what I have seen the latest update was 2 years ago)
3. This pull request is missing the meson build system support - in the meantime we updated HACKING so you should see which of the meson files you should modify.
4. Since the language uses braces, I'd suggest adding it to `lexer_has_braces()` in `editor.c` - see HACKING.

Apart from these suggestions the pull request looks good to me in general so once they are resolved, I think we could merge it.

**Comment by killerbees19** [comment]
Thanks for reviewing my PR, I'll look into all comments in the next few weeks. Currently I'm too busy for this PR :-)

_Just for the record: I'm using this PR since May with patched v1.36 and v1.38 packages at Debian/Ubuntu without any problems. Editing CIL files is much easier than before and I've created thousands of new lines with this syntax highlighter. Anyways, your recommended corrections are highly appreciated and I'll try to fix all issues!_

File: data/filetype_extensions.conf [hunk 6]
```diff
@@ -15,6 +15,7 @@ BibTeX=*.bib;
 CAML=*.ml;*.mli;
 C=*.c;*.xpm;
 C++=*.cpp;*.cxx;*.c++;*.cc;*.h;*.hpp;*.hxx;*.h++;*.hh;*.C;*.H;
+CIL=*.il
```
[the remark this answers is not in the document]
Comment (killerbees19) [reply data/filetype_extensions.conf:18]: Are you sure this is a good idea? It looks similar, but there seem to be significant differences. I don't think SMALI files would benefit from this. Not even the comments are identical.

GitHub syntax highlighter for SMALI:

github-smali

Geany with CIL syntax highlighter:

smali-with-cil-highlighter

File: data/filedefs/filetypes.cil [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+
+[indentation]
+#width=4
+# 0 is spaces, 1 is tabs, 2 is tab & spaces
+#type=1
+
+[build_settings]
+## %f will be replaced by the complete filename
+## %e will be replaced by the filename without extension
+#compiler=ilasm /OUTPUT="%e" "%f"
```
[the remark this answers is not in the document]
Comment (killerbees19) [reply data/filedefs/filetypes.cil:36]: I would assume that it's a standard tool. I don't know if there are (m)any alternatives. In any case, it's mentioned in the official documentation: https://learn.microsoft.com/en-us/dotnet/framework/tools/ilasm-exe-il-assembler

**Comment by killerbees19** [comment]
(Work in progress, just a short _"PR is still alive"_ comment from my side. Hopefully I'll finish and test this PR in august.)

> It would be good to rebase this pull request on top of current master.

Done.

> If there's an updated CIL lexer in Lexilla, please use it (but from what I have seen the latest update was 2 years ago)

No update available.

> This pull request is missing the meson build system support - in the meantime we updated HACKING so you should see which of the meson files you should modify.

`LexCIL.cxx` is already defined in this file?

> Since the language uses braces, I'd suggest adding it to lexer_has_braces() in editor.c - see HACKING.

Done.

File: data/filedefs/filetypes.cil [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+[settings]
```
[the remark this answers is not in the document]
Comment (killerbees19) [reply data/filedefs/filetypes.cil:22]: I don't know what to do at this point. What exactly should I do, I don't understand it right now. Please give me another hint for the right direction. Thank you!

(I have added the MIME type in the meantime.)

[the remark this answers is not in the document]
Comment (techee) [reply data/filedefs/filetypes.cil:22]: > I don't know what to do at this point. What exactly should I do, I don't understand it right now. Please give me another hint for the right direction.

When looking at some other file types like

https://github.com/geany/geany/blob/master/data/filedefs/filetypes.sh

something like this is still missing:

```ini
# these characters define word boundaries when making selections and searching
# using word matching options
#wordchars=_abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789

# single comments, like # in this file
comment_single=#
# multiline comments
#comment_open=
#comment_close=

# set to false if a comment character/string should start at column 0 of a line, true uses any
# indentation of the line, e.g. setting to true causes the following on pressing CTRL+d
# 		#command_example();
# setting to false would generate this
# #		command_example();
# This setting works only for single line comments
comment_use_indent=true

# context action command (please see Geany's main documentation for details)
context_action_cmd=
```

(modify it where needed)

File: data/filetype_extensions.conf (hunk 6 above)
[the remark this answers is not in the document]
Comment (techee) [reply data/filetype_extensions.conf:18]: OK, if there are bigger differences between the languages, it's really better to use CIL just for `*.il` files.

**Comment by techee** [comment]
>> This pull request is missing the meson build system support - in the meantime we updated HACKING so you should see which of the meson files you should modify.

> LexCIL.cxx is already defined in this file?

I'm not sure if I understand what you mean but you have to add the lexer into

https://github.com/geany/geany/blob/71969f600023a333bcd3b717a434082a7039f316/meson.build#L242

**Comment by killerbees19** [comment]
> I'm not sure if I understand what you mean but you have to add the lexer into
> 
> https://github.com/geany/geany/blob/71969f600023a333bcd3b717a434082a7039f316/meson.build#L242

This is already included in this PR since last year. ^^

https://github.com/geany/geany/pull/3480/files#diff-30d8f6be6320feeacf686be94f48c70869b52630e01ea625f0f15adc0d57c3e4R254

**Review by techee — COMMENTED** [review]
> so i think this feature is complete, if we are considering merging it - 

Agree, I think it's mostly ready.

I just had a look at the code once more and found some minor issues but nothing big.

> we should settle on config UI (either keep as is or change to point https://github.com/geany/geany/pull/1 from https://github.com/geany/geany/pull/3911#pullrequestreview-2136554129)
And probably rename everything to "persistent notepad" or something like that.

I've been thinking about it and maybe "persistent untitled documents" would be best so there's no new terminology. I still think it would be best to merge the configuration with "instant save" but depends on what others (@b4n @eht16) think.

Note that we seem to be in the middle of the "Geany summer vacation" period and even normally it takes quite some time to get things merged (so please be patient) but from me at least this PR has full support - it's very useful and it also fixes the issue with the most thumbs up reactions.

**Comment by killerbees19** [comment]
@techee I think your comment should go to PR #3911? 😄

Because it's all Greek to me. 🙂

**Comment by techee** [comment]
>> I've been thinking about it and maybe "persistent untitled documents" ...

> @techee I think your comment should go to PR https://github.com/geany/geany/pull/3911? 😄

> Because it's all Greek to me. 🙂

Good analysis :-).

> This is already included in this PR since last year. ^^

Right, I missed that.

Alright, the PR looks good to me - anything missing from your point of view?

**Review by techee — APPROVED** [review]

**Comment by killerbees19** [comment]
Okay, [final change](https://github.com/geany/geany/compare/27f691cbca5869ac867d80113074facd3ee3b476..c942622e453362f1f93a9d73e143815c5ae26db8): `[build_settings]` commented out, because they are not 100% platform-independent. But I think they are still useful as hint for custom configs.

Successfully built Geany (c942622e453362f1f93a9d73e143815c5ae26db8) and tested it with some CIL files at Debian 12. Looks good to me…

**Comment by techee** [comment]
Looks good to me as well - I'll wait for about one week and if nobody has any objections, I'll merge this PR.

**Comment by techee** [comment]
Merged.

@killerbees19 Thanks for your contribution!

**Pull request merged**